xarray/core/indexes.py

```diff
@@ -30,7 +30,10 @@ def __contains__(self, key):
         return key in self._indexes
 
     def __getitem__(self, key):
-        return self._indexes[key]
+        index = self._indexes[key]
+        # Issue 2949: Pandas indexes have mutable names, but we want to return
+        # immutable views.
+        return index.copy(deep=False)
```

Comment:
Does anyone know if we may face performance issues by making lots of views? (we're never copying the underlying data, but we'll have a `Series` instance on each call)

Reply:
Shallow copies should be pretty cheap. But the real question is whether pandas copies the underlying hash-table. If not, this could cause the performance of repeated indexing to fall off a cliff.

Reply:
Unfortunately, it looks like the hash-table *does not* get copied by `copy(deep=False)`, even though it probably could. Compare:
```
In [11]: import numpy as np

In [12]: import pandas as pd

In [13]: x = np.random.RandomState(0).randint(int(1e6), size=int(1e7))

In [14]: index = pd.Index(x)

# hash-table is not built yet
In [15]: %time index.get_loc(10)
CPU times: user 475 ms, sys: 9.79 ms, total: 485 ms
Wall time: 486 ms
Out[15]: array([False, False, False, ..., False, False, False])

# hash table is cached, so look-up is fast
In [16]: %time index.get_loc(10)
CPU times: user 10.8 ms, sys: 4.75 ms, total: 15.6 ms
Wall time: 14.1 ms
Out[16]: array([False, False, False, ..., False, False, False])

# copied index is slow -- it's rebuilding the hash table!
In [17]: %time index.copy(deep=False).get_loc(10)
CPU times: user 461 ms, sys: 12.8 ms, total: 474 ms
Wall time: 474 ms
Out[17]: array([False, False, False, ..., False, False, False])

# a shallow copy itself is pretty fast
In [18]: %timeit index.copy(deep=False)
16 µs ± 456 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
Ah, good catch. What does this mean for us? Sometimes returning an immutable index (e.g. in `to_series`) and sometimes not (e.g. in `indexes`) seems confusing to me.

Reply:
@shoyer: any more thoughts on this? Happy to see this through or just close.

Reply:
The situation here is a little disappointing, but unfortunately I don't think this fix is workable. I don't sacrificing performance here is worthwhile. It would largely defeat the purpose of indexing in the first place.